### PR TITLE
fix(tokenizer): load merged EOS token IDs from config.json + generation_config.json

### LINKED ISSUE #1112: [Bug]: gRPC TRT-LLM path misses split EOS tokens for Kimi-K2.5, causing overgeneration past turn boundaries
## Bug Description

Kimi-K2.5 defines different EOS tokens in two config files:
- `config.json`: `eos_token_id: 163585` (`[EOS]`)
- `generation_config.json`: `eos_token_id: 163586` (`<|im_end|>`)

The model needs to stop at **both** tokens. In the affected gRPC path, only one EOS token is propagated as `stop_token_ids`, while Kimi-K2.5 requires both `[EOS]` and `<|im_end|>`. SGLang handles this internally by merging both sources, but the TRT-LLM gRPC path does not.

This causes severe overgeneration — the model generates past `<|im_end|>` turn boundaries, producing multi-turn spillover output.

## Steps to Reproduce

**Environment:**
- SMG: `v1.4.1` (commit `5319dc6e`)
- Image: `ghcr.io/lightseekorg/smg:1.4.1-trtllm-1.3.0rc10`
- TRT-LLM: commit `a2557ad9`
- Model: `nvidia/Kimi-K2.5-NVFP4`

**Launch:**
```bash
smg serve \
  --model <path-to-Kimi-K2.5-NVFP4> \
  --router-model-path nvidia/Kimi-K2.5-NVFP4 \
  --router-tool-call-parser kimik2 \
  --router-reasoning-parser kimi_k25 \
  --backend trtllm --tp_size 8 \
  --host 0.0.0.0 --port 8990
```

**Request:**
```bash
curl -s http://localhost:8990/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "nvidia/Kimi-K2.5-NVFP4",
    "messages": [{"role": "user", "content": "What is the weather in San Francisco?"}],
    "tools": [
      {"type": "function", "function": {"name": "getCurrentWeather", "description": "Get the current weather in a given location", "parameters": {"type": "object", "properties": {"location": {"type": "string", "description": "City and state"}}, "required": ["location"]}}}
    ],
    "max_tokens": 500
  }'
```

## Expected Behavior

Model generates one assistant turn (~10 tokens for a single tool call), stops at `<|im_end|>` (163586).

## Actual Behavior

Model generates **253 tokens** for what should be a single tool call. Output is 723 chars of multi-turn spillover — the model generates past `<|im_end|>`, hallucinating a user turn, a tool response, and a follow-up assistant message:

```
content length: 723 chars
completion_tokens: 253 (expected ~10)
leaked markers: ['<|im_end|>', '<|im_assistant|>', '<|im_middle|>']
```

The model doesn't stop at `<|im_end|>` (163586) because only `[EOS]` (163585) is in the gRPC `stop_token_ids`. The `<|im_end|>` token is defined in `generation_config.json` but SMG doesn't merge it into the stop tokens sent to TRT-LLM.

SGLang backend is **not** affected (handles EOS merging internally).

## Component

- Tokenizer (`crates/tokenizer/src/tiktoken.rs`, `crates/tokenizer/src/traits.rs`)
- gRPC Client (`crates/grpc_client/src/trtllm_service.rs`)
- Model Gateway (request building stages)

## Connection Mode

gRPC (TRT-LLM backend only)

## Streaming Context

- [x] Issue occurs with streaming (SSE) responses
- [x] Issue occurs with non-streaming responses
- [x] Issue occurs in both modes

### PR DESCRIPTION
**Pull request by ConnorLi96**
## Problem

Models like Kimi-K2.5 define different EOS tokens in `config.json` (EOS=163585) and `generation_config.json` (`<|im_end|>`=163586). TRT-LLM's gRPC path only uses one source, causing infinite generation past turn boundaries (18/60 → 55/60 on fc-dash).

https://huggingface.co/nvidia/Kimi-K2.5-NVFP4/blob/main/config.json#L12
https://huggingface.co/nvidia/Kimi-K2.5-NVFP4/blob/main/generation_config.json#L3

## Solution

- Add `load_eos_token_ids()` to merge EOS IDs from both config files
- Expose `eos_token_ids()` on the `Tokenizer` trait
- Pass merged IDs as `stop_token_ids` in TRT-LLM gRPC requests
- Remove the hardcoded `<|im_end|>` stop string workaround

## Changes

- `crates/tokenizer/src/traits.rs` — add `eos_token_ids()` to trait
- `crates/tokenizer/src/tiktoken.rs` — `load_eos_token_ids()` impl
- `crates/tokenizer/src/huggingface.rs` — impl for HF tokenizer
- `crates/tokenizer/src/mock.rs` — mock impl
- `crates/grpc_client/src/trtllm_service.rs` — pass merged IDs, remove hardcoded stop string
- `model_gateway/src/routers/grpc/client.rs` — plumb new field
- `model_gateway/.../harmony/stages/request_building.rs` — pass EOS IDs
- `model_gateway/.../regular/stages/chat/request_building.rs` — pass EOS IDs

## Test Plan

- Verified Kimi-K2.5 on TRT-LLM: fc-dash 55/60 (was 18/60 without EOS fix)
- SGLang path unaffected (handles EOS merging internally)

Checklist:
- [x] `cargo +nightly fmt` passes
- [x] `cargo clippy --all-targets --all-features -- -D warnings` passes

Made with [Cursor](https://cursor.com)

## Summary by CodeRabbit

* **New Features**
  * Support for discovering and loading End-of-Sequence (EOS) token IDs from model config files.
  * Tokenizer trait and implementations now expose EOS token IDs and propagate them through caching wrappers.
  * Request builders and request construction paths now accept and forward explicit EOS token ID data.
  * Tokenizer download logic expanded to include additional config files and a new public EOS helper module.

* **Tests**
  * Added unit tests covering EOS ID extraction, merging, and edge cases.

Closes #1112

### CODE REVIEW

File: crates/tokenizer/src/huggingface.rs [hunk 1]
```diff
@@ -82,6 +82,12 @@ impl HuggingFaceTokenizer {
             }
         }
 
+        // Load merged EOS token IDs from config.json + generation_config.json
+        let mut special_tokens = special_tokens;
+        if let Some(dir) = std::path::Path::new(file_path).parent() {
```
[the remark this answers is not in the document]
Comment (slin1237) [reply crates/tokenizer/src/huggingface.rs:87]: this is not a nit
this is a critical piece that should be fixed

File: crates/tokenizer/src/huggingface.rs [hunk 2]
```diff
@@ -82,6 +82,12 @@ impl HuggingFaceTokenizer {
             }
         }
 
+        // Load merged EOS token IDs from config.json + generation_config.json
+        let mut special_tokens = special_tokens;
```
Comment (slin1237) [inline crates/tokenizer/src/huggingface.rs:86]: when loading hf or tiktokenizer
do we not have those special tokens already?

File: crates/tokenizer/src/tiktoken.rs [hunk 3]
```diff
@@ -42,6 +42,60 @@ struct TiktokenConfig {
...
+                            ids.insert(id as TokenIdType);
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+
+    ids.into_iter().collect()
```
[the remark this answers is not in the document]
Comment (slin1237) [reply crates/tokenizer/src/tiktoken.rs]: this should not be nit
it should be addressed before merging

File: crates/grpc_client/src/trtllm_service.rs [hunk 4]
```diff
@@ -273,6 +277,7 @@ impl TrtllmServiceClient {
         token_ids: Vec<u32>,
         multimodal_input: Option<proto::MultimodalInput>,
         tool_call_constraint: Option<(String, String)>, // (constraint_type, constraint_value)
+        eos_token_ids: &[u32],
```
Comment (slin1237) [inline crates/grpc_client/src/trtllm_service.rs:280]: why do we only change trt?
and is this specific to kimi-k2?

**Review by slin1237 — CHANGES_REQUESTED** [review]

File: model_gateway/src/routers/grpc/harmony/stages/request_building.rs [hunk 5]
```diff
@@ -175,6 +175,10 @@ impl PipelineStage for HarmonyRequestBuildingStage {
                 ProtoGenerateRequest::Vllm(Box::new(req))
             }
             GrpcClient::Trtllm(trtllm_client) => {
+                let eos_ids = ctx
```
Comment (slin1237) [inline model_gateway/src/routers/grpc/harmony/stages/request_building.rs]: this is not even supported for harmony afaik

File: crates/grpc_client/src/trtllm_service.rs (hunk 4 above)
Comment (ConnorLi96) [reply crates/grpc_client/src/trtllm_service.rs:280]: since SGLang and vLLM merge EOS tokens internally — they read both config files in their Python startup code. so only TRTLLM. 

not Kimi-specific, the fix is generic, just Kimi-K2.5 is the first model we see this issue.

File: crates/tokenizer/src/huggingface.rs (hunk 2 above)
Comment (ConnorLi96) [reply crates/tokenizer/src/huggingface.rs:86]: `tokenizer_config.json` only provides `eos_token` as a string (e.g. "[EOS]"), but not the token ID. The actual EOS token IDs come from `config.json` and `generation_config.json`. Again, SGLang and vLLM handle this merge internally in their Python code. TRT-LLM's gRPC path doesn't, so SMG needs to read and merge them explicitly.

Also found and fixed a related issue: when SMG loads the tokenizer via HF Hub (by model ID rather than local path), the hub downloader only fetched tokenizer files, not `config.json/generation_config.json`. Fixed in hub.rs so these are now included in the download.

File: crates/tokenizer/src/huggingface.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ConnorLi96) [reply crates/tokenizer/src/huggingface.rs:87]: Agreed, fixed. Extracted `collect_eos_ids()` helper and replaced the duplication with a loop over both config filenames. Also moved the function out of `tiktoken.rs` into a shared crates/tokenizer/src/eos.rs module.

File: model_gateway/src/routers/grpc/harmony/stages/request_building.rs (hunk 5 above)
Comment (ConnorLi96) [reply model_gateway/src/routers/grpc/harmony/stages/request_building.rs]: You're right, Harmony doesn't run on TRT-LLM. the TRT-LLM arm just passes `&[]` now to satisfy the function signature.

since we change the signature of `build_generate_request_from_chat` and `build_generate_request_from_responses`, we have to do this otherwise can't compile IIUC.

File: crates/grpc_client/src/trtllm_service.rs (hunk 4 above)
Comment (CatherineSue) [reply crates/grpc_client/src/trtllm_service.rs:280]: Fix this in trt instead. It is not in our scope to pass this to their backend, just because this is a bug in their end.

**Comment by uralik** [comment]
@slin1237  could you take a look on the latest update here? this is quite huge issue that prevents seamlessly using Kimi models served via vllm with smg. thanks!
